[ruby] Build/test ruby 3.4 and build native gems with Ruby 3.4 support

chadlwilson opened the pull request:
Updates ruby-compiler-dock to `1.7.0` to [bring Ruby 3.4 support](https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.7.0) and starts cross-compiling native gems for Ruby 3.4. Added a distribtest as well.

I believe from last year at #35399 that this probably requires @apolcyn or another Googler to build+push+update the `grpc-testing` images at
https://github.com/grpc/grpc/blob/17715928ad96b0a99041ed7658599e36d4f98d30/tools/bazelify_tests/dockerimage_current_versions.bzl#L22-L29

chadlwilson commented on Rakefile:
```diff
@@ -143,7 +143,7 @@ task 'gem:native', [:plat] do |t, args|
 
   grpc_config = ENV['GRPC_CONFIG'] || 'opt'
-  ruby_cc_versions = ['3.3.0', '3.2.0', '3.1.0', '3.0.0'].join(':')
+  ruby_cc_versions = ['3.4.0', '3.3.5', '3.2.0', '3.1.0', '3.0.0'].join(':')
```
Changing to `3.3.5` is necessary due to the notes at https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.6.0 (Ruby `3.3.0` had an issue with cross-compilation only fixed in  `3.3.5`)

apolcyn commented:
Thanks @chadlwilson !

I've uploaded docker image changes here and sent you https://github.com/chadlwilson/grpc/pull/5

I think tests should pass after merging https://github.com/chadlwilson/grpc/pull/5

chadlwilson commented:
Thanks @apolcyn! I realised I forgot to push the generated `Dockerfile` for the distribtest image with 3.4 on Debian 11, now added.

Are you able to help build+push that one too?

apolcyn commented:
@chadlwilson let's try https://github.com/chadlwilson/grpc/pull/7

chadlwilson commented:
I think the distribtests should be good now, if someone could kindly nudge the tests again :-)

chadlwilson commented:
Unfortunately, the distribtests are back to `ruby_abi_version` errors on Linux

```
+ bundle exec ./distribtest.rb
bundler: failed to load command: ./distribtest.rb (./distribtest.rb)
/usr/local/rvm/rubies/ruby-3.4.0/lib/ruby/3.4.0+1/bundled_gems.rb:82:in 'Kernel.require': /usr/local/rvm/gems/ruby-3.4.0/gems/grpc-1.70.0.dev-x86_64-linux/src/ruby/lib/grpc/3.4/grpc_c.so: undefined symbol: ruby_abi_version - ruby_abi_version (LoadError)
	from /usr/local/rvm/rubies/ruby-3.4.0/lib/ruby/3.4.0+1/bundled_gems.rb:82:in 'block (2 levels) in Kernel#replace_require'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/grpc-1.70.0.dev-x86_64-linux/src/ruby/lib/grpc/grpc.rb:20:in '<top (required)>'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/grpc-1.70.0.dev-x86_64-linux/src/ruby/lib/grpc.rb:19:in 'Kernel#require_relative'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/grpc-1.70.0.dev-x86_64-linux/src/ruby/lib/grpc.rb:19:in '<top (required)>'
	from /usr/local/rvm/rubies/ruby-3.4.0/lib/ruby/3.4.0+1/bundled_gems.rb:82:in 'Kernel.require'
	from /usr/local/rvm/rubies/ruby-3.4.0/lib/ruby/3.4.0+1/bundled_gems.rb:82:in 'block (2 levels) in Kernel#replace_require'
	from /var/local/git/grpc/test/distrib/ruby/distribtest.rb:17:in '<top (required)>'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/cli/exec.rb:59:in 'Kernel.load'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/cli/exec.rb:59:in 'Bundler::CLI::Exec#kernel_load'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/cli/exec.rb:23:in 'Bundler::CLI::Exec#run'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/cli.rb:452:in 'Bundler::CLI#exec'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/vendor/thor/lib/thor/command.rb:28:in 'Bundler::Thor::Command#run'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in 'Bundler::Thor::Invocation#invoke_command'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/vendor/thor/lib/thor.rb:538:in 'Bundler::Thor.dispatch'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/cli.rb:35:in 'Bundler::CLI.dispatch'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/vendor/thor/lib/thor/base.rb:584:in 'Bundler::Thor::Base::ClassMethods#start'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/cli.rb:29:in 'Bundler::CLI.start'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/exe/bundle:28:in 'block in <top (required)>'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/lib/bundler/friendly_errors.rb:117:in 'Bundler.with_friendly_errors'
	from /usr/local/rvm/gems/ruby-3.4.0/gems/bundler-2.6.2/exe/bundle:20:in '<top (required)>'
	from /usr/local/rvm/gems/ruby-3.4.0/bin/bundle:25:in 'Kernel#load'
	from /usr/local/rvm/gems/ruby-3.4.0/bin/bundle:25:in '<main>'
	from /usr/local/rvm/gems/ruby-3.4.0/bin/ruby_executable_hooks:22:in 'Kernel#eval'
	from /usr/local/rvm/gems/ruby-3.4.0/bin/ruby_executable_hooks:22:in '<main>'
```

This looks related to the discussion/fix at #31970 / #31151 / #30976 - not sure if something went awry with the symbol export/stripping?

flavorjones commented:
https://github.com/rake-compiler/rake-compiler-dock/pull/142 should fix the ruby_abi_version issue.

flavorjones commented:
Please update to use https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.7.1

apolcyn commented:
@chadlwilson regarding rake-compiler-dock 1.7.1, let's try https://github.com/chadlwilson/grpc/pull/8

chadlwilson commented:
Thanks @apolcyn - merged now, so we can give the tests another go any time. :-)

chadlwilson commented:
The `ruby_abi_version` issues continued with the updated rake-compiler-dock, so presumably the issue must be with use of `3.4.0` at gem install time rather than cross-compile time.

I've tried updating the distribtest ruby version to `3.4.1` in https://github.com/grpc/grpc/pull/38338/commits/84b1953f40bdd26dd00fa58fb089b761eca589f3 (and updated the default local/unit test version for consistency) 

@apolcyn would you be able to rebuild/push the `distribtest/ruby_debian11_x64_ruby_3_4` image again when you get a chance?

apolcyn commented:
> [The ruby_abi_version issues continued](https://btx.cloud.google.com/invocations/14875588-3546-4b46-a9e3-cab4a97a4e8e/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_distribtests_ruby;config=default/log) with the updated rake-compiler-dock, so presumably the issue must be with use of 3.4.0 at gem install time rather than cross-compile time.

Hmm, it would be unfortunate if our pre-compiled gems didn't work on ruby 3.4.0, though.

I'd like to understand the root cause here a bit better, and only declare ruby 3.4.0 incompatible as a last resort.

I.e. I think we need to figure out a fix that allows this to work on ruby 3.4.0 - so let's keep the distribtest ruby runtime as is.

chadlwilson commented:
@apolcyn Yeah, I hear you - I don't fully understand the problem either, however it's quite clear that the ruby `3.4.0` release had a seemingly important bug here (hence the immediate patch release) that seems likely to cause an issue in one configuration or another since they tagged `RUBY_PATCHLEVEL` incorrectly as a development build for that release. One might argue that `3.4.0` is a bad release that should be ignored? (although the Ruby team dont seem to have specifically said that, or what the issue might be expected to break).

I requested updating the test images initially as I find it very difficult to reproduce this locally since
- gems seem to work fine via `3.4.0` with a direct extension build on both darwin arm64 and a linux aarch64 container (**without** `rake-compiler-dock` and cross-compile scripts)
- `rake-compiler-dock` images are only x64 compiled (https://github.com/rake-compiler/rake-compiler-dock/issues/78). My main decent spec machine is aarch64 hardware. The boringssl compiles seg fault under QEMU within Docker so running rake-compiler-dock cross compiles is tough for me.
- I can't see the rake-compiler-dock logs in Kokoro to confirm which flags it ran make with for various versions.

That's mainly why I was hoping to see if we can get to a working configuration of the distribtests, and then decide how/whether to proceed - since I dont seem to have an environment to replicate locally. If you're open to trying this (just to validate our understanding) before considering merging anything, that'd help - but your call.

**Re: the actual problem**

**Logically speaking,** if Ruby is "looking for" the `ruby_abi_version` symbol before loading the extension, it seems to think that the Ruby runtime is a development build and thus needs to check abi version compatibility for extensions. That seems to make sense that it'd possibly fail if running with `3.4.0` (bad patch level) but extension was compiled with `3.4.1` and so extension didn't have/export symbol.

However it doesn't seem to explain why it failed even when the extension was compiled with `3.4.0` as in the original configuration as naively I'd expect `have_ruby_abi_version()` to have been true, and it to have included the symbol (not stripped it).

What seemed _possible_ to me to explain the original failure (`3.4.0` test version, `3.4.0` cross-compile target), is that when cross-compiling it was actually making these decisions based on the rake-compiler-dock bootstrap version (`3.1.3p185`) rather than the target version - since I believe the cross-compiles really just changes includes, but the `extconf` otherwise runs under the bootstrap version's Ruby. So the logic below would still consider it to be a proper release version `have_ruby_abi_version() = false`.

https://github.com/grpc/grpc/blob/260b41b9cdcf361d4a4c12c3bcfd7294e5ca4b18/src/ruby/ext/grpc/extconf.rb#L149-L152

This would mean that the symbols wouldn't be exported and so potentially when the gem tries to be installed with `3.4.0` it is mistakenly trying to check the abi version and thus failing.

If that theory is accurate, I'd guess that `3.4.0` is a not going to work with cross-compiled gems with the current code.

Other info from upstream:
- it looks like [in `3.2.1` release](https://github.com/ruby/ruby/releases/tag/v3_2_1) they changed it to not strip the symbol for release versions via https://bugs.ruby-lang.org/issues/19289 so the patch added in https://github.com/grpc/grpc/pull/31997 might not be needed in the same way and we can change things to just keep the symbol for all Ruby versions (except `3.2.0`, unsure about truffleruby) and perhaps remove the `RUBY_PATCHLEVEL` check entirely.

chadlwilson commented:
Still failed with the [same error here](https://btx.cloud.google.com/invocations/0d595b04-f7d8-410b-af58-99031ca7392c/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_distribtests_ruby;config=default/log).

```
bundler: failed to load command: ./distribtest.rb (./distribtest.rb)
/usr/local/rvm/rubies/ruby-3.4.0/lib/ruby/3.4.0+1/bundled_gems.rb:82:in 'Kernel.require': /usr/local/rvm/gems/ruby-3.4.0/gems/grpc-1.70.0.dev-x86_64-linux/src/ruby/lib/grpc/3.4/grpc_c.so: undefined symbol: ruby_abi_version - ruby_abi_version (LoadError)
```

So I suppose we can scratch my theories.

adomokos commented:
I tried using Ruby 3.4 for my projects, but that failed fairly quickly. Can you just make grpc work for 3.4.1 and skip 3.4?

chadlwilson commented:
> I tried using Ruby 3.4 for my projects, but that failed fairly quickly. Can you just make grpc work for 3.4.1 and skip 3.4?

In what way did 3.4.0 fail for your projects? And 3.4.1 worked?

Right now we don't *know* 3.4.1.works any better with the cross-compiled binaries than 3.4.0 (we haven't tried it here yet, at least).

I do have a (slowish) x64 machine available to me so if no-one has better ideas/abilities I'll try replicating the issue on that where I might have a chance of getting the docker-based cross-compile to complete, so I can then run the distrib tests myself.

apolcyn commented:
@chadlwilson also just to experiment I've pushed your updates to distribtest images: https://github.com/chadlwilson/grpc/pull/9

chadlwilson commented:
I've merged that now, so we can try the tests again. I'll try and get some time to set things up on my other x64 machine in the next day or two.

chadlwilson commented:
Well, running the distribtests with `3.4.1` [allowed the Ruby Linux tests to pass](https://source.cloud.google.com/results/invocations/13ef4cb3-3c33-49a9-8379-0bdc3c6388f2) - so we have a way forward if we are OK to ignore 3.4.0. The other CI failures are unrelated and appear to be flaky tests; will likely pass with a re-run.

I guess to move this forward and address @apolcyn 's concerns we should
- better understand why the cross-compiled gems are failing when installed to `3.4.0` when normal gem makes are fine.
- evaluate whether a workaround is possible to allow them to work on `3.4.0` or whether we should just ignore `3.4.0` as a "bad Ruby release" and support only 3.4.1+
- arguably sanity test the cross-compiled MacOS artifacts as well, as the distribtests don't validate MacOS right now.
- decide after that whether removing the earlier `RUBY_PATCHLEVEL` workaround in [this commit](https://github.com/grpc/grpc/pull/38338/commits/9d7f0e9fa045689294216c44d39091056d5b74e1) is still a good idea to reduce coupling between ruby releases and grpc ruby builds.

apolcyn commented:
@chadlwilson let's try https://github.com/chadlwilson/grpc/pull/10

My read of what's happening is this:

- ruby 3.4.0 accidentally had `RUBY_PATCHLEVEL` set to -1 (https://github.com/ruby/ruby/compare/v3_4_0...v3_4_1). Meanwhile `rb_abi_version` was left unset.
  - normally, *in release* builds, I guess `RUBY_PATCHLEVEL` should be >= 0 (and `rb_abi_version` should be unset). Note `RUBY_PATCHLEVEL` being -1 normally implies that we're on a dev ruby version and that `rb_abi_version` exists.

So I think a quick fix will be to just to special case ruby 3.4.0 in `have_ruby_abi_version` - have hardcode that to return `false`.

apolcyn commented:
@chadlwilson sorry typo mistake: `../../../../src/ruby/ext/grpc/extconf.rb:270: syntax error, unexpected end-of-input, expecting `end'`

if we can fix that I can give it another run

chadlwilson commented:
@apolcyn fixed now

chadlwilson commented:
[Still failed on 3.4.0](https://btx.cloud.google.com/invocations/93c98841-d04d-4ce6-b9c8-265569506682/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_distribtests_ruby;config=default/log) in a similar inexplicable way to my experiments above.

I believe because the decision made in the extconf is relevant only at (cross-)compile time, not runtime which could be across different 3.4.x versions for the same gem.

>   * normally, _in release_ builds, I guess `RUBY_PATCHLEVEL` should be >= 0 (and `rb_abi_version` should be unset). Note `RUBY_PATCHLEVEL` being -1 normally implies that we're on a dev ruby version and that `rb_abi_version` exists.

What I tried to outline above when removing the patchlevel hack is that supposedly Ruby 3.2.1+ was changed to include rb_abi_version in all builds https://bugs.ruby-lang.org/issues/19289 specifically referring to avoiding brittle code in grpc.and differences between dev and release ruby versions (but here we are anyway with patchlevel problems). But maybe I misunderstood it as my earlier theories didn't work.

I'll keep digging, but I still don't see why we should care about 3.4.0. It was available for *less than 2 hours* before 3.4.1 was released - presumably released with such urgency because the numbering problem is so fatal?

chadlwilson commented:
Sure, have merged.

In terms of iterating locally I have an x64 env I can `rake-compiler-dock` running now, but my x64 linux compile timed out after 2 hours (probably filesystem mount speed with colima on Mac) so need to re-run :D

> Yes I'm also not quite sure what's going on here (and how that [patch](https://github.com/ruby/ruby/commit/6f3aff3961a4c5ce87e05096a1a9dcf1055b7647#diff-c5f1fb44ef5931d015ac302543368ea882ccdf7424b043ec1394903094db4844L37) relates)

Yes, that patch was a bit strange to me also.

> My main concern with 3.4.0 support is if this issue might be _specific_ to grpc (since the end result will be more grpc-specific headaches). I'm still willing to _not_ support it in that case, I'm just trying to see if we can fix this easily enough.

Yes, it is a bit odd that I can't find anyone else observing such issues with `3.4.0` (and it works fine when you compile it directly via extconf), so figured it either related to grpc's custom symbol stripping/retaining logic or style of use of `rake-compiler-dock` somehow, but the failure of the latest attempt seems to even call the former into question!

dazuma commented:
Just a note that rake-compiler-dock 1.8.0 is available. I doubt it affects anything related to Ruby 3.4.0, but we will be using that version to build google-protobuf with Ruby 3.4 support.

stanhu commented:
Note that rack-compiler-dock 1.7.1 updated to Ruby 3.4.1 to fix the `ruby_abi_version` issue (https://github.com/rake-compiler/rake-compiler-dock/issues/141): https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.7.1

chadlwilson commented:
> Note that rack-compiler-dock 1.7.1 updated to Ruby 3.4.1 to fix the `ruby_abi_version` issue ([rake-compiler/rake-compiler-dock#141](https://github.com/rake-compiler/rake-compiler-dock/issues/141)): https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.7.1

@stanhu Yes, we already updated to that in https://github.com/grpc/grpc/pull/38338/commits/7a2b11741eae0949cd2eb4db6d4aadbf2ddd7d7b (it was partially on my request in https://github.com/rake-compiler/rake-compiler-dock/issues/141 ) - didn't fix the problem.

apolcyn commented:
@chadlwilson since this isn't regressing on support for ruby 3.4.0 (currently pre-compiled grpc will fail to load on any 3.4 version), I'll propose we go ahead with this change  (leaving 3.4.0 as something to fix separately, if possible).

chadlwilson commented:
Thanks @apolcyn - I've rebased off master and removed the experimental `3.4.0` commits, leaving this on rake-compiler-dock `1.7.1`, building locally and testing with ruby `3.4.1`.

Do you want me to 
- update to [rake-compiler-dock 1.8.0](https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.8.0) in this PR for sake of getting to latest version incl for cross-compiling previous ruby feature versions, or not worry about that for now?
- squash commits further?

apolcyn commented:
> update to [rake-compiler-dock 1.8.0](https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.8.0) in this PR for sake of getting to latest version incl for cross-compiling previous ruby feature versions, or not worry about that for now?

Yeah this looks good to have -  let's do this, while we're here.

Don't worry about rebasing (that will happen automatically when this gets merged)

dazuma commented:
> > update to [rake-compiler-dock 1.8.0](https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.8.0) in this PR for sake of getting to latest version incl for cross-compiling previous ruby feature versions, or not worry about that for now?
> 
> Yeah this looks good to have - let's do this, while we're here.

The one gotcha I ran into when doing this for protobuf is that `RUBY_CC_VERSION` for rake-compiler-dock 1.8.0 are (near-)latest patchlevels of each Ruby minor version, as listed in the release notes, rather than the .0 patch levels used in 1.7 and earlier. Just in case it might save a few minutes of head-scratching. 😁

chadlwilson commented:
> The one gotcha I ran into when doing this for protobuf is that `RUBY_CC_VERSION` for rake-compiler-dock 1.8.0 are (near-)latest patchlevels of each Ruby minor version, as listed in the release notes, rather than the .0 patch levels used in 1.7 and earlier. Just in case it might save a few minutes of head-scratching. 😁

Ack, yup - similar to the 1.7.1 update.

> > update to [rake-compiler-dock 1.8.0](https://github.com/rake-compiler/rake-compiler-dock/releases/tag/v1.8.0) in this PR for sake of getting to latest version incl for cross-compiling previous ruby feature versions, or not worry about that for now?
> 
> Yeah this looks good to have - let's do this, while we're here.

@apolcyn I've updated now in https://github.com/grpc/grpc/pull/38338/commits/b8060ba6b5d0355d38cb892a98b14e36c7e869e9 - if you could help build and push new images when you get a chance.

apolcyn commented:
>  if you could help build and push new images when you get a chance.

done, just sent https://github.com/chadlwilson/grpc/pull/13

chadlwilson commented:
Looks like build/tests have passed, so think we are looking good here :-)

chadlwilson commented:
For anyone watching this, this was reverted in #38516 (`master`) and #38515 (from `1.70`) due to some internal build issues that it seems the community cannot easily figure out right now.
